File: public/codes/2-Regresion/6-Random_Forest_Regression/random_forest_regression.R
# =============================================================================
# Random Forest Regression
# =============================================================================

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================
# Estructura de los datos:tipos de empleados y el nivel 
# de cada tipo de empleado y el salario correspondiente.
# Objetivo: asignar un salario correspondientes a un nivel y su posicion dada.
# Filas :10
# Columnas:
#           |Position|Level| (vars independiente)
#           |salary| (var_dependiente)
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]

# No se hace ninguna distincion entre variables independientes 
# y variables dependientes en R

# =============================================================================
# Ajustar la regresion random forest con el dataset
# =============================================================================
# crear nuestra variable de regresion aqui
# library(randomForest)
set.seed(1)
rf_regressor=randomForest(x=dataset[1],y=dataset$Salary,
                          ntree=1000)
summary(rf_regressor)

# =============================================================================
# Visualizacion de los resultado: Modelo random forest
# =============================================================================
# Para mostrar la grafica de datos de entrenamiento

x_grid=seq(min(dataset$Level),max(dataset$Level),0.01)
# Agregando componentes a mostrar
ggplot()+
  # Dobujando los puntos de entrenamiento
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             color="red")+
  # Dobujando la linea de la prediccion, en base al entrenamiento
  # geom_line(aes(x=dataset$Level,y=predict(rf_regressor,newdata=dataset)),
            geom_line(aes(x=x_grid,
                          y=predict(rf_regressor,
                                    newdata=data.frame(Level=x_grid))),
            color="blue")+
  ggtitle("Prediccion usando regresion con arboles aleatorios de decision   ")+
  xlab("labelx")+
  ylab("labely")

# =============================================================================
# Prediccion de nuestros modelos (Resultados)
# =============================================================================
y_pred_tree=predict(rf_regressor,newdata=data.frame(Level=6.5))
print(y_pred_tree)

File: public/codes/2-Regresion/6-Random_Forest_Regression/random_forest_regression.py
# -*- coding: utf-8 -*-
"""
Date: Wed Feb 22 00:08:15 2023

@author: edwin
"""
# =============================================================================
# Random Forest Regression
# =============================================================================

# =============================================================================
# --------------------Importando librerias--------------------
# =============================================================================
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================

dataset = pd.read_csv('Position_Salaries.csv') # {buscar el dataset}

# Variable independiente:Mayuscula por ser una matriz.
X = dataset.iloc[:,1:2].values

# Variable dependiente:minuscula por ser un vector.
y = dataset.iloc[:,2:3].values

# =============================================================================
# Ajustar la regresion con arboles aleatorios de decision con el dataset
# =============================================================================

from sklearn.ensemble import RandomForestRegressor 
rf_regression=RandomForestRegressor(n_estimators=1000,random_state=0)
rf_regression.fit(X,y)
y_pred=rf_regression.predict(X)
# =============================================================================
# Visualizacion de los resultado: Random Forest
# =============================================================================

X_grid =np.arange(min(X),max(X),0.1)
X_grid=X_grid.reshape(len(X_grid),1)

plt.scatter(X,y,color='red')
plt.plot(X_grid,rf_regression.predict(X_grid),color='green')
plt.plot(X,rf_regression.predict(X),color='green')
plt.title("Modelo Regresion por arboles aleatorios de decision")
plt.xlabel("Posicion del empleado")
plt.ylabel("Sueldo en $")
plt.show()

# =============================================================================
# Prediccion de nuestros modelos (Resultados)
# =============================================================================
y_pred=rf_regression.predict([[6.5]])
print(y_pred)

